hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/qjournal/client/TestQuorumJournalManager.java

```diff
@@ -1101,6 +1113,59 @@ public void testSelectViaRpcTwoJNsError() throws Exception {
     }
   }
 
+  /**
+   * Test selecting EditLogInputStream after some journalNode jitter.
+   * And the corner case as below:
+   * 1. Journal 0 has some abnormal cases when journaling Edits with start txId 11.
+   * 2. NameNode just ignore the abnormal journal 0 and continue to write Edits to Journal 1 and 2.
+   * 3. Journal 0 backed to health.
+   * 4. Observer NameNode try to select EditLogInputStream vis PRC with start txId 21.
+   * 5. Journal 1 has some abnormal cases caused slow response.
+   *
+   * And the expected selecting result is: Response should contain 20 Edits from txId 21 to txId 40.
+   * Because there is no Edits from id 21 to 40 in the cache of JournalNode0.
+   */
+  @Test
+  public void testSelectViaRpcAfterJNJitter() throws Exception {
+    EditLogOutputStream stm = qjm.startLogSegment(
+        1, NameNodeLayoutVersion.CURRENT_LAYOUT_VERSION);
+    SettableFuture<Void> slowLog = SettableFuture.create();
+    Mockito.doReturn(slowLog).when(spies.get(0))
+        .sendEdits(eq(1L), eq(11L), eq(1), Mockito.any());
+    writeTxns(stm, 1, 10);
+    writeTxns(stm, 11, 10);
+    writeTxns(stm, 21, 10);
+    writeTxns(stm, 31, 10);
+    ListeningExecutorService service = MoreExecutors.listeningDecorator(
+        Executors.newSingleThreadExecutor());
+    Mockito.doAnswer(invocation -> service.submit(
+        () -> {
+          ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+          EditLogFileOutputStream.writeHeader(
+              NameNodeLayoutVersion.CURRENT_LAYOUT_VERSION,
+              new DataOutputStream(byteStream));
+          byteStream.write(createTxnData(21, 20));
+          Thread.sleep(3000);
+          return GetJournaledEditsResponseProto.newBuilder()
+              .setTxnCount(20)
+              .setEditLog(ByteString.copyFrom(byteStream.toByteArray()))
+              .build();
+        })
+    ).when(spies.get(1)).getJournaledEdits(21,
+        QuorumJournalManager.QJM_RPC_MAX_TXNS_DEFAULT);
```

Comment:
Why don't we just use `invocation.callRealMethod()` instead of constructing the response here?

```suggestion
    Mockito.doAnswer(invocation -> service.submit(() -> {
      Thread.sleep(3000);
      try {
        return invocation.callRealMethod();
      } catch (Throwable e) {
        throw (Exception) e;
      }
    })).when(spies.get(1)).getJournaledEdits(21,
        QuorumJournalManager.QJM_RPC_MAX_TXNS_DEFAULT);
```

Reply:
Nice suggestion. Thanks, Sir.